src/app/loadout-builder/mod-assignments.ts

```diff
@@ -0,0 +1,140 @@
+import { DestinyEnergyType } from 'bungie-api-ts/destiny2';
+import _ from 'lodash';
+import { DimItem, PluggableInventoryItemDefinition } from '../inventory/item-types';
+import { ProcessItem, ProcessMod } from './process-worker/types';
+
+const DEFAULT_ENERGY_INVESTMENT = 1000;
+
+interface ItemAssignments {
+  mods: ProcessMod[];
+  energySwapped: boolean;
+}
+
+export class ModAssignments {
+  // We just keep track of the id's and hashes as we assume only valid items and mod
+  // combninations are passed in
+  private slotDependantAssignments: Map<string, number[]>;
+  // We need to keep track of a bit more data here so we track the item and the data
+  // kept in ItemAssignments
+  private slotIndependantAssignments: Map<ProcessItem, ItemAssignments>;
+  // Keep track of the current set of assignments extra energy investment so we dont
+  // need to continually recalculate it
+  private extraEnergyInvestment = DEFAULT_ENERGY_INVESTMENT;
+
+  constructor() {
+    this.slotDependantAssignments = new Map();
+    this.slotIndependantAssignments = new Map();
+  }
+
+  get isSuccessfullyAssigned() {
+    return this.extraEnergyInvestment < DEFAULT_ENERGY_INVESTMENT;
+  }
+
+  /**
+   * This returns a map of item id's to mod hashes for the assigned mods.
+   */
+  getResults() {
+    const rtn = new Map<string, number[]>();
+
+    for (const [item, itemAssignments] of this.slotIndependantAssignments.entries()) {
+      rtn.set(
+        item.id,
+        itemAssignments.mods.map((mod) => mod.hash)
+      );
+    }
+
+    for (const [itemId, mods] of this.slotDependantAssignments.entries()) {
+      if (rtn.has(itemId)) {
+        rtn.get(itemId)?.push(...mods);
+      } else {
+        rtn.set(itemId, mods);
+      }
+    }
+
+    return rtn;
+  }
+
+  /**
+   * Assigns slot dependant mods (i.e. helmet mods) to the item.
+   */
+  assignSlotDependantMods(item: DimItem, mods: PluggableInventoryItemDefinition[]) {
+    this.slotDependantAssignments.set(
+      item.id,
+      mods.map((mod) => mod.hash)
+    );
+  }
+
+  // Calculates the energy used and wasted for an item and its assignments.
+  private calculateEnergyChange(item: ProcessItem, assignments: ItemAssignments) {
```

Comment:
I thought we actually had a lint for this, but this can be a free function as it doesn't use `this`.